config/v1/types_infrastructure.go

```diff
@@ -505,9 +505,113 @@ type BareMetalPlatformStatus struct {
 	NodeDNSIP string `json:"nodeDNSIP,omitempty"`
 }
 
+type OpenStackAPIBGPPeer struct {
+	// asn is the ASN of the peer.
+	//
+	// +kubebuilder:validation:Required
+	ASN string `json:"asn"`
+
+	// ip is the IP address of the peer. It may be either IPv4 or IPv6.
+	//
+	// +kubebuilder:validation:Required
+	IP string `json:"ip"`
+}
+
+// OpenStackAPIBGPSpeaker species
+type OpenStackAPIBGPSpeaker struct {
+	// failureDomain is the name of a failure domain which this BGP
+	// configuration applies to.
+	//
+	// +kubebuilder:validation:Required
+	FailureDomain string `json:"failureDomain"`
+
+	// asn specifies the ASN to be used by the BGP speaker.
+	//
+	// +kubebuilder:validation:Required
+	ASN string `json:"asn"`
+
+	// peers is a list of all BGP peers of this speaker.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	Peers []OpenStackAPIBGPPeer `json:"peers"`
+}
+
+type OpenStackAPIBGPConfiguration struct {
+	// apiVIPs is a list of IP addresses which will be published via BGP for
+	// use by the API server. It may contain both IPv4 and IPv6 addresses.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	APIVIPs []string `json:"apiVIPs"`
+
+	// speakers is a list of BGP speaker configurations. We require a
+	// speaker configuration for every failure domain hosting a control
+	// plane node.
+	//
+	// +kubebuilder:validation:MinItems:=1
+	// +listType=map
+	// +listMapKey=failureDomain
+	Speakers []OpenStackAPIBGPSpeaker `json:"speakers"`
+}
+
+// OpenStackFailureDomain specifies a failure domain for an OpenStack server.
+// Specifically, it specifies a set of values which will be set on all machines
+// using the failure domain.
+type OpenStackFailureDomain struct {
+	// name is the unique name of this failure domain.
```

Comment:
Any length restrictions or pattern this must adhere to?
Must this relate to anything in OSP or is it arbitrary?
Is this an OpenShift specific ID and if so, why does it exist, what is it used for

Reply:
This name is specific to OpenShift, although in practise it will probably usually correspond to the name of an OpenStack availability zone.

It will be used to refer to the failure domain in the Machine definition, which isn't included here.

In the context of CPMS, it will also be used to determine the failure domain of a machine. Because we're defining a FD to be a set of values, reverse engineering the FD from the OpenStack server isn't straightforward. Consequently I'm currently expecting it to be explicit in the machine spec, as well as having all its values substituted in the spec at creation time.

Reply:
added in the docstring:
* it's arbitrary
* it is used in the BGP configuration